owltorrent-network/src/main/java/edu/rice/owltorrent/network/messages/PieceMessage.java

```diff
@@ -0,0 +1,65 @@
+package edu.rice.owltorrent.network.messages;
+
+import edu.rice.owltorrent.common.entity.FilePiece;
+import edu.rice.owltorrent.common.entity.Torrent;
+import edu.rice.owltorrent.network.PeerMessage;
+import java.nio.ByteBuffer;
+import lombok.EqualsAndHashCode;
+import lombok.Value;
+
+/**
+ * Piece Message used to send/receive piece messages.
+ *
+ * @author Max Yu
+ */
+@Value
+@EqualsAndHashCode(callSuper = true)
+public class PieceMessage extends PeerMessage {
+
+  /** 2 * 4 from the index and begin and one index */
+  private static final int BASE_SIZE = 9;
+
+  int index;
+  int begin;
+  byte[] piece;
+
+  public PieceMessage(int index, int begin, byte[] piece) {
+    super(MessageType.PIECE);
+    this.index = index;
+    this.begin = begin;
+    this.piece = piece;
+  }
+
+  public FilePiece getFilePiece() {
+    // FIXME: why is this hash in filePiece? no data?
+    return new FilePiece(index, begin, piece);
+  }
+
+  @Override
+  public byte[] toBytes() {
+    ByteBuffer buffer = ByteBuffer.allocate(4 + PieceMessage.BASE_SIZE + piece.length);
+    buffer.putInt(PieceMessage.BASE_SIZE + piece.length);
+    buffer.put(MessageType.PIECE.getByteValue());
+    buffer.putInt(index);
+    buffer.putInt(begin);
+    buffer.put(piece);
+
+    // Sets position to 0
+    buffer.rewind();
+    return buffer.array();
+  }
+
+  @Override
+  public boolean verify(Torrent torrent) {
```

Comment:
Only the piece message should have a verify method (doesn't make sense for e.g. the KeepAliveMessage to). Also this is just my opinion but I think this should be done upon construction in the parse method (so the parse method takes in a Torrent file as well).